src/Lecture4.hs

```diff
@@ -183,9 +212,18 @@ The 'Stats' data type has multiple fields. All these fields have
 instance for the 'Stats' type itself.
 -}
 
-instance Semigroup Stats where
 
 
+instance Semigroup Stats where
+  (<>) :: Stats -> Stats -> Stats
+  (Stats a1 b1 c1 d1 e1 f1 g1 h1 j1) <> (Stats a2 b2 c2 d2 e2 f2 g2 h2 j2) =
+    (Stats (a1 <> a2) (b1 <> b2) (c1 <> c2) (d1 <> d2) (e1 `eagerMaybeAppend` e2) (f1 `eagerMaybeAppend` f2) (g1 `eagerMaybeAppend` g2) (h1 `eagerMaybeAppend` h2) (j1 <> j2))
+    where 
+      eagerMaybeAppend Nothing Nothing     = Nothing 
+      eagerMaybeAppend Nothing (Just !x)   = Just x
+      eagerMaybeAppend (Just !x) Nothing   = Just x 
+      eagerMaybeAppend (Just !x) (Just !y) = Just (x <> y) 
```

Comment:
That's absolutely correct implementation 👍🏻 
You can write it a bit shorter though by pattern matching only on the final result and the implementation will be lazier in a "better" way 🙂 

```suggestion
      eagerMaybeAppend a b = case a <> b of
          Nothing -> Nothing
          Just !x -> Just x 
```